src/smt_util/node_traversal.h

```diff
@@ -0,0 +1,117 @@
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Iterators for traversing nodes.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__SMT_UTIL__NODE_TRAVERSAL_H
+#define CVC4__SMT_UTIL__NODE_TRAVERSAL_H
+
+#include <cstddef>
+#include <iterator>
+#include <vector>
+
+#include "expr/node.h"
+
+namespace CVC4 {
+
+// Iterator for traversing a node in post-order
+// It does DAG-traversal, so indentical sub-nodes will be visited once only.
+class NodePostorderIterator
+{
+ public:
+  // Construct a traversal iterator beginning at `n`
+  NodePostorderIterator(TNode n);
+  // Construct an end-of-traversal iterator
+  NodePostorderIterator();
+
+  // STL type definitions for an iterator
+  using value_type = TNode;
+  using pointer = TNode*;
+  using reference = TNode&;
+  using iterator_category = std::forward_iterator_tag;
+  using difference_type = std::ptrdiff_t;
+
+  // Move/copy construction and assignment. Destructor.
+  NodePostorderIterator(NodePostorderIterator&&) = default;
+  NodePostorderIterator& operator=(NodePostorderIterator&&) = default;
+  NodePostorderIterator(NodePostorderIterator&) = default;
+  NodePostorderIterator& operator=(NodePostorderIterator&) = default;
+  ~NodePostorderIterator() = default;
+
+  // Preincrement
+  NodePostorderIterator& operator++();
+  // Postincrement
+  NodePostorderIterator operator++(int);
+  // Dereference
+  reference operator*();
+  // Equals
+  bool operator==(const NodePostorderIterator&) const;
+  // Not equals
+  bool operator!=(const NodePostorderIterator&) const;
+
+ private:
+  // General Invariant: The top node on the stack (`d_stack.back()`) is the
+  // current location of the traversal.
+  //
+  // There is one exception to this: when an iterator is constructed
+  // (the stack has one node and nothing has been visited)
+  //
+  // While we could expand the stack, adding children until the top node is a
+  // leaf, we do not do so, because this is expensive, and we want construction
+  // to be cheap.
+  std::vector<TNode> d_stack;
+
+  // Whether (and how) we've visited a node.
+  // Absent if we haven't visited it.
+  // Set to `false` if we've pre-visited it (enqueued its children).
+  // Set to `true` if we've also post-visited it.
+  std::unordered_map<TNode, bool, TNodeHashFunction> d_visited;
+
+  // Continues the traversal until the next post-visit.
+  void advanceToNextPostVisit();
+
+  // Return `true` iff this iterator has not been dereferenced or incremented
+  // yet.
+  bool justConstructed() const;
+};
+
+// Node wrapper that is iterable in DAG post-order
+class NodePostorderIterable
+{
+ public:
+  NodePostorderIterable(TNode n);
+
+  // STL type definitions for an iterable
+  using iterator = NodePostorderIterator;
+  using value_type = TNode;
+  using reference = TNode&;
+  using difference_type = std::ptrdiff_t;
```

Comment:
Also define the `iterator_category`. I believe that this is a `forward_iterator_tag`?

Reply:
Hmm. Since this isn't actually an iterator (it is a range-expression), I thought that defining `iterator_category` would be inappropriate. Am I mistaken?

My goal was to support range-based for loops. Now that I look at the [requirements](https://en.cppreference.com/w/cpp/language/range-for) though, I don't believe I even need these declarations. How about I get rid of all of them?

Reply:
Oh right, yes. We should only define them on the `NodePostorderIterator` itself and not the `NodePostorderIterable`. I got the classes mixed up.